File: solutions/cpp/5.longest-palindromic-substring.cpp
#include <iostream>

using namespace std;

// Manacher's Algorithm
class Solution {
public:
    string longestPalindrome(string s) {
        string t = "^#";
        for (int i = 0; i < s.length(); ++i) {
            t += s[i];
            t += '#';
        }
        int p[t.length()], id = 0, mx = 0, resid = 0, resmx = 0;
        memset(p, 0, sizeof p);
        for (int i = 1; i < t.length(); ++i) {
            p[i] = mx > i ? min(p[2 * id - i], mx - i) : 1;
            while (t[i + p[i]] == t[i - p[i]]) ++p[i];
            if (mx < i + p[i]) {
                mx = i + p[i];
                id = i;
            }
            if (resmx < p[i]) {
                resmx = p[i];
                resid = i;
            }
        }
        return s.substr((resid - resmx) / 2, resmx - 1);
    }
};

/* Dynamic Programming
 * dp[i][j] = 1 if j == i
 * dp[i][j] = (s[i] == s[j]) if j = i + 1
 * dp[i][j] = ((s[i] == s[j] && dp[i + 1][j-1]) if j > i + 1 */
class SolutionDP {
public:
    string longestPalindrome(string s) {
        if (s.empty()) return "";
        int n = s.length(), dp[n][n], left = 0, len = 1;
        memset(dp, 0, sizeof dp);
        for (int i = 0; i < n; ++i) {
            dp[i][i] = 1;
            for (int j = 0; j < i; ++j) {
                dp[j][i] = (s[i] == s[j] && (i - j < 2 || dp[j + 1][i - 1]));
                if (dp[j][i] && len < i - j + 1) {
                    len = i - j + 1;
                    left = j;
                }
            }
        }
        return s.substr(left, len);
    }
};

int main() {
    Solution soln;
    string s = "babad";
    string res = soln.longestPalindrome(s);
    cout << "s: " << s << endl;
    cout << "result: " << res << endl;
    return 0;
}

File: solutions/python/5.longest-palindromic-substring.py
# 5. Longest Palindromic Substring

# Dynamic Programming
class Solution_dp:
    def longestPalindrome(self, s: str) -> str:
        longest = 0
        l = len(s)
        if l == 1 or l == 0: return s
        if l == 2:
            if s[0] == s[1]: return s
            else: return s[0]
        dp = [None] * l
        if s[0] == s[1]:
            dp[:2] = [True] * 2
            longest = 2
            ip, jp = 1, 0
        else:
            dp[0], dp[1] = False, True
            longest = 1
            ip, jp = 0, 0
        for i in range(2, l):
            for j in range(i+1):
                if i - j > 1:
                    dp[j] = s[j] == s[i] and dp[j+1]
                    if dp[j]:
                        palindromic = i - j + 1
                        if longest < palindromic:
                            longest = palindromic
                            ip, jp = i, j
                if i - j == 1:
                    dp[j] = s[j] == s[i]
                    if longest < 2 and dp[j]:
                        longest = 2
                        ip, jp = i, j
                if i - j == 0:
                    dp[j] = True
        return s[jp:ip+1]


# Palindromic Substring Equals Its Reverse
class Solution(object):
    def longestPalindrome(self, s: str) -> str:
        if len(s)<2 or s==s[::-1]:
            return s
        n=len(s)
        start,maxlen=0,1
        for i in range(n):
            odd =s[i-maxlen-1:i+1]
            even=s[i-maxlen:i+1]
            if i-maxlen-1>=0 and odd==odd[::-1]:
                start=i-maxlen-1
                maxlen+=2
                continue
            if i-maxlen>=0 and even==even[::-1]:
                start=i-maxlen
                maxlen+=1
        return s[start:start+maxlen]


def test():
    s = Solution()

    string = "babad"
    output = s.longestPalindrome(string)
    print(f'{string=}\n{output=}\n')

    string = "cbbd"
    output = s.longestPalindrome(string)
    print(f'{string=}\n{output=}\n')

    string = ""
    output = s.longestPalindrome(string)
    print(f'{string=}\n{output=}\n')

    string = "ccc"
    output = s.longestPalindrome(string)
    print(f'{string=}\n{output=}\n')

    string = "ccd"
    output = s.longestPalindrome(string)
    print(f'{string=}\n{output=}\n')

    string = "caba"
    output = s.longestPalindrome(string)
    print(f'{string=}\n{output=}')


if __name__ == '__main__':
    test()
